[Filebeat] Add postgresql csv fileset

azlev opened the pull request:
As a configuration option, create another pipeline inside the module to
parse CSV log files.

Although it's necessary to configure the database to emit such logs,
there is some advantages: configured properly, some events like
`statement timeout` and `lock timeout` will display the query in the
same event, opposed to multiple lines, one with the error message and
other with statement.

## What does this PR do?

This PR creates a new fileset/pipeline inside PostgreSQL module. This new fileset parses and create events based on CSV log files from postgresql.

## Why is it important?

PostgreSQL logs weren't designed to be parsed, so, without very complicated logic, it's not possible to encapsulate all information in just one event. One example of that is `statement timeout`:

```
2021-01-04 02:07:15.470 UTC [13470] [user=postgres,db=postgres,app=[unknown]] ERROR:  canceling statement due to statement timeout
2021-01-04 02:07:15.470 UTC [13470] [user=postgres,db=postgres,app=[unknown]] STATEMENT:  SELECT pg_sleep(100)
```
Using the CSV log, those 2 lines become 1:
```
2021-01-04 02:07:15.470 UTC,"postgres","postgres",13470,"127.21.1.7:28182",5ff27834.349e,1,"SELECT",2021-01-04 02:06:44 UTC,16/4181895,0,ERROR,57014,"canceling statement due to statement timeout",,,,,,"SELECT pg_sleep(100)",,,""
```

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] Documentation

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23334/detail/PR-23334/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23334/detail/PR-23334/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23334/detail/PR-23334/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23334/detail/PR-23334/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `Pull request #23334 opened`

* Reason: `The PR is not allowed to run in the CI yet`

* Start Time: 2021-01-04T02:37:25.092+0000

* Duration: 4 min 37 sec


* Commit: 684f912ce00fe6768bf9f5d0fc32f9c4010c42a3




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->





### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23334/detail/PR-23334/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-23334/runs/1/steps/83/log/?start=0">here</a></li>
<li>Description: <code>githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-01-04T02:39:06.805Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-04T02:39:06.808Z]  > git merge b498aa3527f38f1aa0e210b6d8ff51be60eeaf28 # timeout=10
[2021-01-04T02:39:08.055Z] Merge succeeded, producing 91cc1ac50e45573b0a8fccb9d35c2db44fa36937
[2021-01-04T02:39:08.055Z] Checking out Revision 91cc1ac50e45573b0a8fccb9d35c2db44fa36937 (PR-23334)
[2021-01-04T02:39:08.048Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-01-04T02:39:08.057Z]  > git config core.sparsecheckout # timeout=10
[2021-01-04T02:39:08.062Z]  > git checkout -f 91cc1ac50e45573b0a8fccb9d35c2db44fa36937 # timeout=15
[2021-01-04T02:39:11.678Z] Commit message: "Merge commit 'b498aa3527f38f1aa0e210b6d8ff51be60eeaf28' into HEAD"
[2021-01-04T02:39:11.678Z] First time build. Skipping changelog.
[2021-01-04T02:39:11.678Z] Cleaning workspace
[2021-01-04T02:39:11.840Z]  > git --version # timeout=10
[2021-01-04T02:39:11.843Z]  > git --version # 'git version 2.17.1'
[2021-01-04T02:39:11.846Z] fatal: bad object 7edba8e3544d4f6700385847f4f4c4daad2361f5
[2021-01-04T02:39:11.680Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-04T02:39:11.684Z] Resetting working tree
[2021-01-04T02:39:11.684Z]  > git reset --hard # timeout=10
[2021-01-04T02:39:11.771Z]  > git clean -fdx # timeout=10
[2021-01-04T02:39:12.804Z] Timeout set to expire in 3 hr 0 min
[2021-01-04T02:39:12.824Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2021-01-04T02:39:13.035Z] [INFO] Number of builds to be searched 10
[2021-01-04T02:39:13.618Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2021-01-04T02:39:13.638Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-23334/src/github.com/elastic/beats
[2021-01-04T02:39:13.666Z] [INFO] gitCheckout: Checkout SCM PR-23334 with default customisation from the Item.
[2021-01-04T02:39:13.695Z] [INFO] Override default checkout
[2021-01-04T02:39:13.741Z] Sleeping for 10 sec
[2021-01-04T02:39:23.767Z] The recommended git tool is: git
[2021-01-04T02:39:23.811Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2021-01-04T02:39:23.839Z] Wiping out workspace first.
[2021-01-04T02:39:23.847Z] Cloning the remote Git repository
[2021-01-04T02:39:23.847Z] Using shallow clone with depth 10
[2021-01-04T02:39:23.847Z] Avoid fetching tags
[2021-01-04T02:39:23.858Z] Cloning repository git@github.com:elastic/beats.git
[2021-01-04T02:39:23.878Z]  > git init /var/lib/jenkins/workspace/Beats_beats_PR-23334/src/github.com/elastic/beats # timeout=10
[2021-01-04T02:39:23.883Z] Fetching upstream changes from git@github.com:elastic/beats.git
[2021-01-04T02:39:23.883Z]  > git --version # timeout=10
[2021-01-04T02:39:23.888Z]  > git --version # 'git version 2.17.1'
[2021-01-04T02:39:23.888Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-04T02:39:23.892Z]  > git fetch --no-tags --progress -- git@github.com:elastic/beats.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2021-01-04T02:39:42.197Z] Cleaning workspace
[2021-01-04T02:39:42.212Z] Using shallow fetch with depth 10
[2021-01-04T02:39:42.212Z] Pruning obsolete local branches
[2021-01-04T02:39:42.183Z]  > git config remote.origin.url git@github.com:elastic/beats.git # timeout=10
[2021-01-04T02:39:42.186Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-01-04T02:39:42.193Z]  > git config remote.origin.url git@github.com:elastic/beats.git # timeout=10
[2021-01-04T02:39:42.199Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-04T02:39:42.206Z] No valid HEAD. Skipping the resetting
[2021-01-04T02:39:42.206Z]  > git clean -fdx # timeout=10
[2021-01-04T02:39:42.214Z] Fetching upstream changes from git@github.com:elastic/beats.git
[2021-01-04T02:39:42.215Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-04T02:39:42.218Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/beats.git +refs/pull/23334/head:refs/remotes/origin/PR-23334 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2021-01-04T02:39:43.369Z] Merging remotes/origin/master commit b498aa3527f38f1aa0e210b6d8ff51be60eeaf28 into PR head commit 684f912ce00fe6768bf9f5d0fc32f9c4010c42a3
[2021-01-04T02:39:43.371Z]  > git config core.sparsecheckout # timeout=10
[2021-01-04T02:39:43.377Z]  > git checkout -f 684f912ce00fe6768bf9f5d0fc32f9c4010c42a3 # timeout=15
[2021-01-04T02:39:44.808Z]  > git remote # timeout=10
[2021-01-04T02:39:44.811Z]  > git config --get remote.origin.url # timeout=10
[2021-01-04T02:39:44.815Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-01-04T02:39:44.818Z]  > git merge b498aa3527f38f1aa0e210b6d8ff51be60eeaf28 # timeout=10
[2021-01-04T02:39:46.218Z] Merge succeeded, producing e014d49416736af1ce464367a658243c594da80c
[2021-01-04T02:39:46.218Z] Checking out Revision e014d49416736af1ce464367a658243c594da80c (PR-23334)
[2021-01-04T02:39:46.300Z] Commit message: "Merge commit 'b498aa3527f38f1aa0e210b6d8ff51be60eeaf28' into HEAD"
[2021-01-04T02:39:46.300Z] First time build. Skipping changelog.
[2021-01-04T02:39:46.300Z] Cleaning workspace
[2021-01-04T02:39:46.460Z]  > git --version # timeout=10
[2021-01-04T02:39:46.463Z]  > git --version # 'git version 2.17.1'
[2021-01-04T02:39:46.466Z] fatal: bad object 7edba8e3544d4f6700385847f4f4c4daad2361f5
[2021-01-04T02:39:46.214Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-01-04T02:39:46.221Z]  > git config core.sparsecheckout # timeout=10
[2021-01-04T02:39:46.223Z]  > git checkout -f e014d49416736af1ce464367a658243c594da80c # timeout=15
[2021-01-04T02:39:46.302Z]  > git rev-parse --verify HEAD # timeout=10
[2021-01-04T02:39:46.305Z] Resetting working tree
[2021-01-04T02:39:46.305Z]  > git reset --hard # timeout=10
[2021-01-04T02:39:46.395Z]  > git clean -fdx # timeout=10
[2021-01-04T02:39:47.440Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2021-01-04T02:39:48.045Z] + git fetch https://****:****@github.com/elastic/beats.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2021-01-04T02:40:55.816Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-23334/src/github.com/elastic/beats/.git
[2021-01-04T02:40:55.847Z] Archiving artifacts
[2021-01-04T02:40:56.536Z] + git rev-parse HEAD
[2021-01-04T02:40:56.858Z] + git rev-parse HEAD
[2021-01-04T02:40:57.161Z] + git rev-parse origin/pr/23334
[2021-01-04T02:40:57.205Z] [INFO] githubEnv: Found Git Build Cause: pr
[2021-01-04T02:40:57.498Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-01-04T02:40:58.722Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/beats/pulls/23334/reviews return 0 elements
[2021-01-04T02:40:58.770Z] [INFO] githubPrCheckApproved: Title: [Filebeat] Add postgresql csv fileset - User: azlev - Author Association: FIRST_TIME_CONTRIBUTOR
[2021-01-04T02:40:59.067Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2021-01-04T02:40:59.067Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2021-01-04T02:40:59.133Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2021-01-04T02:40:59.153Z] Sleeping for 5 sec
[2021-01-04T02:41:00.726Z] Stage "Lint" skipped due to earlier failure(s)
[2021-01-04T02:41:00.786Z] Stage "Build&Test" skipped due to earlier failure(s)
[2021-01-04T02:41:00.847Z] Stage "Packaging" skipped due to earlier failure(s)
[2021-01-04T02:41:00.930Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-23334/src/github.com/elastic/beats
[2021-01-04T02:41:01.356Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-23334
[2021-01-04T02:41:01.591Z] [INFO] getVaultSecret: Getting secrets
[2021-01-04T02:41:01.653Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-01-04T02:41:02.407Z] + chmod 755 generate-build-data.sh
[2021-01-04T02:41:02.407Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23334/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23334/runs/1 ABORTED 217049
[2021-01-04T02:41:02.658Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23334/runs/1/steps/?limit=10000 -o steps-info.json
[2021-01-04T02:41:03.208Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23334/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-01-04T02:41:03.208Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-01-04T02:41:04.551Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

jsoriano commented on filebeat/module/postgresql/csv/_meta/fields.yml:
```diff
@@ -0,0 +1,71 @@
+        Possible types are autovacuum launcher, autovacuum worker, logical replication launcher,
+        logical replication worker, parallel worker, background writer, client backend, checkpointer,
+        startup, walreceiver, walsender and walwriter.
+        In addition, background workers registered by extensions may have additional types.
```
There is no need to add these fields, they are already documented in the `log` metricset. We could consider moving their definition to the module-level `fields.yml`, but by now, let's just avoid duplicating them.

I don't know though if we have another fileset that collects the same fields as other fileset.


```
```

azlev replied:
I'm in doubt here. Not all fields are defined in the postgresql.log fileset.

So, should I move duplicates to beats/filebeat/module/postgresql/_meta/fields.yml` and leave others inside the fileset? Is there any example around the code?

jsoriano replied:
Ok, I found an example in the Kafka module in Metricbeat, there we have some partition fields that are filled by multiple metricsets, they are defined at the module level, here:
https://github.com/elastic/beats/blob/8e8bfc8786abded0bca05da7e7cbb8751b96061b/metricbeat/module/kafka/_meta/fields.yml#L38-L51

With filebeat modules it should work the same, we can put common fields at the module level.

azlev replied:
Since I merged `csv` inside `log`, this part is solved.

jsoriano commented on filebeat/module/postgresql/csv/_meta/fields.yml:
```diff
@@ -0,0 +1,71 @@
+- name: log
```
This is going to generate duplicated field groups, this name should be unique in the module.
```suggestion
- name: csv
```

jsoriano replied:
Let's explicitly mark it as beta by now.
```suggestion
- name: csv
  release: beta
```

jsoriano commented on filebeat/module/postgresql/_meta/docs.asciidoc:
````diff
@@ -68,4 +73,31 @@ image::./images/filebeat-postgresql-slowlog-overview.png[]
+log_disconnections = on;
+log_lock_waits = on;
+log_min_duration_statement = 0;
+```
````
:+1:

jsoriano reviewed (commented):
Thanks for this contribution, very interesting, and quite complete!

jsoriano commented:
jenkins run the tests please

jsoriano commented:
jenkins run the tests please

jsoriano commented:
By the way, there is another approach to support multiple formats. For example the elasticsearch module contains pipelines for plain and JSON logs, and it chooses automatically which one to use.

Multiple pipelines can be included like this in the fileset manifest:
https://github.com/elastic/beats/blob/8e8bfc8786abded0bca05da7e7cbb8751b96061b/filebeat/module/elasticsearch/server/manifest.yml#L15-L18

The first pipeline is executed when an event is received, in this case `pipeline.yml`. It contains some common processing for all formats, but in an specific part, it chooses to run one of the other pipelines depending on the format of the message:
https://github.com/elastic/beats/blob/8e8bfc8786abded0bca05da7e7cbb8751b96061b/filebeat/module/elasticsearch/server/ingest/pipeline.yml#L15-L20

This approach is also used in other modules like kibana, logstash, coredns, envoyproxy... you can look for examples that use the `IngestPipeline` helper in the pipeline.

If we can do something similar for these postgresql logs, we could add a CSV-specific pipeline to the existing `log` metricset, and we could more easily reuse the common parts of the pipelines, as well as the common fields. And it would be easier for users, as they wouldn't need to configure a different fileset depending on the format.

@azlev what do you think? would it be possible to do it this way in this case?

azlev commented:
> By the way, there is another approach to support multiple formats. For example the elasticsearch module contains pipelines for plain and JSON logs, and it chooses automatically which one to use.
> 
> Multiple pipelines can be included like this in the fileset manifest:
> https://github.com/elastic/beats/blob/8e8bfc8786abded0bca05da7e7cbb8751b96061b/filebeat/module/elasticsearch/server/manifest.yml#L15-L18
> 
> The first pipeline is executed when an event is received, in this case `pipeline.yml`. It contains some common processing for all formats, but in an specific part, it chooses to run one of the other pipelines depending on the format of the message:
> https://github.com/elastic/beats/blob/8e8bfc8786abded0bca05da7e7cbb8751b96061b/filebeat/module/elasticsearch/server/ingest/pipeline.yml#L15-L20
> 
> This approach is also used in other modules like kibana, logstash, coredns, envoyproxy... you can look for examples that use the `IngestPipeline` helper in the pipeline.
> 
> If we can do something similar for these postgresql logs, we could add a CSV-specific pipeline to the existing `log` metricset, and we could more easily reuse the common parts of the pipelines, as well as the common fields. And it would be easier for users, as they wouldn't need to configure a different fileset depending on the format.
> 
> @azlev what do you think? would it be possible to do it this way in this case?

I tried a minimal approach to avoid breaking changes, and I assumed that it would be easier to merge. So now I'll try to do a more integrated code. I'll give a try, but those changes will take a while to implement.

azlev commented:
@jsoriano I found 2 ways to choose the ingest pipeline:
* file extension (PostgreSQL always use .csv to those files)
* regular expression like this: ctx =~ /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}(\.\d{1,3})( \w{3,4}),/

I couldn't find any way to get the file name, and this will break the tests. The regular expression doesn't look as a good idea in a critical path like this, and looks fragile.

Do you have other ideas?

jsoriano commented:
> @jsoriano I found 2 ways to choose the ingest pipeline:
>   * file extension (PostgreSQL always use .csv to those files)
>   * regular expression like this: ctx =~ /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}(.\d{1,3})( \w{3,4}),/
> 
> I couldn't find any way to get the file name, and this will break the tests. The regular expression doesn't look as a good idea in a critical path like this, and looks fragile.
> 
> Do you have other ideas?

@azlev I agree that the simpler we can do the selection, the better, to avoid affecting the critical path.

Checking the file extension could be a good option. File path should be available in `log.file.path`. You don't see it in tests because [it is removed](https://github.com/elastic/beats/blob/1327562b17cf06c430a2ad668ba22a3f836606a4/filebeat/tests/system/test_modules.py#L291) to avoid issues validating the expected events, but at the moment of processing the path should be there.
The problem I see with this approach is that not all inputs collect from files, e.g. the postgresql module could be used with logs collected with [syslog](https://www.elastic.co/guide/en/beats/filebeat/7.10/filebeat-input-syslog.html), or from [s3 buckets](https://www.elastic.co/guide/en/beats/filebeat/7.10/filebeat-input-s3.html). So even if this is more straightforward, it may not work in all cases.

Ideally, the decision of using one or the other pipeline should be based on something available in the message itself, as you are doing with the regular expression.

Other possible approach could be to add a `format` setting to the fileset, this value could be set in the events with `add_fields`, so pipelines can decide what to do. But it would be better if the module can do the parsing automatically without depending on user configuration.

Another approach you could try is to parse as grok with the current pipeline, but if grok fails, fallback to the CSV pipeline with [`on_failure`](https://www.elastic.co/guide/en/elasticsearch/reference/master/handling-failure-in-pipelines.html). This can also affect the performance of the pipeline, but maybe you could give it a try if other options are not good enough.

If we are not convinced by any of these approaches I am ok with going back to the approach of using two different filesets.

azlev commented:
I've split the grok pipeline in 2 parts: the first one parse the timestamp plus 1 char. This char will be used to choose the correct pipeline: `log` or `csv`.

In the process, I deprecated the `core_id` field since I couldn't find it inside postgresql documentation.

Still, there is some fields that I need work on, the 2 that I'm currently thinking about are `error.code` and `event.type`

* The PostgreSQL error code is not a number: https://www.postgresql.org/docs/current/errcodes-appendix.html
* I duplicated `event.type` in `error_severity`. I'm checking how to deal with it

azlev reviewed (commented):
@jsoriano I have a question here: https://github.com/elastic/beats/blob/master/filebeat/module/postgresql/log/ingest/pipeline.yml#L40

In the current code, if there is an error, the final event will have: `"event.type": ["info", "error"]`. Is that right?

jsoriano commented on filebeat/docs/fields.asciidoc:
```diff
@@ -105593,7 +105593,49 @@ The timestamp from the log line.
+
+deprecated:[8.0.0]
+
+Core id. (deprecated, there is no core_id in PostgreSQL logs, this is actually session_line_number)
```
Interesting, good catch!

jsoriano commented on filebeat/module/postgresql/log/_meta/fields.yml:
```diff
@@ -8,17 +8,32 @@
         The timestamp from the log line.
     - name: core_id
+      type: alias
+      path: session_line_number
```
Full path is needed here.
```suggestion
      path: postgresql.log.session_line_number
```

jsoriano commented on filebeat/module/postgresql/log/_meta/fields.yml:
```diff
@@ -28,28 +43,72 @@
+        State code returned by Postgres (if any).
+        This code is not a number.
+        See also https://www.postgresql.org/docs/current/errcodes-appendix.html
+    - name: detail
+    - name: hint
```
No types or docs for these ones?

jsoriano commented on filebeat/module/postgresql/log/_meta/fields.yml:
```diff
@@ -28,28 +43,72 @@
+        In addition, background workers registered by extensions may have additional types.
 
     - name: error.code
       type: long
```
Change to `keyword` in the meantime to avoid errors?

Or stop collecting this and make it an alias to `sql_state_code`?

jsoriano commented on filebeat/module/postgresql/log/ingest/pipeline.yml:
```diff
@@ -3,18 +3,24 @@ processors:
-    ignore_missing: true
     patterns:
-    - '^%{DATETIME:postgresql.log.timestamp} (\[%{NUMBER:process.pid:long}(-%{BASE16FLOAT:postgresql.log.core_id:long})?\] ((\[%{USERNAME:user.name}\]@\[%{POSTGRESQL_DB_NAME:postgresql.log.database}\]|%{USERNAME:user.name}@%{POSTGRESQL_DB_NAME:postgresql.log.database}) )?)?%{WORD:log.level}:  (?:%{NUMBER:postgresql.log.error.code:long}|%{SPACE})(duration: %{NUMBER:temp.duration:float} ms  %{POSTG ...
+    - '^%{DATETIME:postgresql.log.timestamp}%{CHAR:separator}%{GREEDYDATA:message2}'
```
Nit. Use a more meaningful name for `message2`, e.g.
```suggestion
    - '^%{DATETIME:postgresql.log.timestamp}%{CHAR:separator}%{GREEDYDATA:raw_message}'
```

jsoriano commented on filebeat/module/postgresql/log/manifest.yml:
```diff
@@ -4,10 +4,16 @@ var:
   - name: paths
     default:
       - /var/log/postgresql/postgresql-*-*.log*
+      - /var/log/postgresql/postgresql-*-*.csv
```
Add a wildcard if these files can be rotated.
```suggestion
      - /var/log/postgresql/postgresql-*-*.csv*
```

jsoriano reviewed (commented):
This is looking great, nice revamp of the module :+1: 

> @jsoriano I have a question here: https://github.com/elastic/beats/blob/master/filebeat/module/postgresql/log/ingest/pipeline.yml#L40
> 
> In the current code, if there is an error, the final event will have: `"event.type": ["info", "error"]`. Is that right?

It doesn't look right. I don't see a reason to have both error types. I think this should be changed to put one or the other, but not both. You can make the change in this PR, but as this is an existing issue feel free to leave this for a future change.

jsoriano commented:
jenkins run the tests please

azlev commented:
I think that I've addressed all previous issues / conversations. Let me know if there is anything else that needs to be changed.

jsoriano commented:
jenkins run the tests please

jsoriano commented on filebeat/module/postgresql/log/_meta/fields.yml:
```diff
@@ -28,25 +43,74 @@
+    - name: transaction_id
+      type: long
+      description: The id of current transaction.
+    - name: sql_state_code
+      description: >
+        State code returned by Postgres (if any).
+        This code is not a number.
+        See also https://www.postgresql.org/docs/current/errcodes-appendix.html
```
Nit. Perhaps it is not needed to show in the public docs that this is not a number, reference to postgres docs is probably enough, we can keep the sentence in a comment here so this is not changed to a numeric type.
```suggestion
    - name: sql_state_code
      # This code is not a number.
      type: keyword
      description: >
        State code returned by Postgres (if any).
        See also https://www.postgresql.org/docs/current/errcodes-appendix.html
```

jsoriano commented on filebeat/module/postgresql/log/ingest/pipeline-csv.yml:
```diff
@@ -0,0 +1,78 @@
+    - '^duration: %{NUMBER:temp.duration:float} ms$'
+    - '^duration: %{NUMBER:temp.duration:float} ms  %{POSTGRESQL_QUERY_STEP} %{DATA:postgresql.log.query_name}: %{GREEDYDATA:message}$'
+    - '^duration: %{NUMBER:temp.duration:float} ms  %{POSTGRESQL_QUERY_STEP}: %{GREEDYDATA:message}$'
+    - '^(%{POSTGRESQL_QUERY_STEP}: )?%{GREEDYDATA:message}$'
```
All patterns seem to require a duration or a query step, query steps here are a list of specific values, if a different query step appears, the processor will fail. Would it be possible to add a pattern, or modify `POSTGRESQL_QUERY_STEP`  to catch this case?

jsoriano replied:
Ah, yes, you are right, I missed this `?`. All good then :+1:

jsoriano commented on filebeat/module/postgresql/log/ingest/pipeline-csv.yml:
```diff
@@ -0,0 +1,78 @@
+      GREEDYDATA: |-
+        (.|
+        |   )*
+      POSTGRESQL_QUERY_STEP: '(parse|bind|statement|fastpath function call|execute|execute fetch from)'
```
Query step is captured in `postgresql.log.query_step` in the plain text pipeline, should it be also captured in this pipeline?

azlev replied:
That one I was in doubt. This `query_step` overlaps with `command_tag`, but not 100%, so I opted to let it empty since I was creating a new fileset at first. Now that I merged, I can review this. I'll populate it, at least to some backward compatibility.

jsoriano commented on filebeat/module/postgresql/log/ingest/pipeline.yml:
```diff
@@ -41,15 +47,22 @@ processors:
     value:
       - info
+    if: "ctx?.postgresql?.log?.sql_state_code == null || (ctx.postgresql.log.sql_state_code ==~ /^0[012].*/)"
 - append:
```
The [`set` processor](https://www.elastic.co/guide/en/elasticsearch/reference/7.11/set-processor.html) could be used to add `event.type` instead of `append`, so it is not set as an array in the document.

jsoriano reviewed (commented):
@azlev I think this is almost ready to merge, I have added some small questions/suggestions, let me know what you think.

jsoriano commented:
jenkins run the tests again please

jsoriano commented on filebeat/module/postgresql/log/test/postgresql-13-error-code.csv.log-expected.json:
```diff
@@ -0,0 +1,145 @@
+        "event.timezone": "UTC",
+        "event.type": [
+            "info"
+        ],
```
We may need to regenerate these files after the change from `append` to `set`.

azlev replied:
Yeah, thats odd, I noted that too, for some reason they didn't failed

azlev replied:
```
 make update && GENERATE=1 INTEGRATION_TESTS=1 BEAT_STRICT_PERMS=false TESTING_FILEBEAT_MODULES=postgresql pytest tests/system/test_modules.py
```
This command didn't do the trick. :confused:

jsoriano replied:
Running tests locally I see this error that seems related:
```
Invalid [path] value [postresql.log.sql_state_code] for field alias [postgresql.log.error.code]: an alias must refer to an existing field in the mappings.
```

Also, a unexpected collateral effect, tests in `filebeat/scripts/tester/main_test.go` have started to fail because they use the postgresql pipelines as data files :grimacing: Could you update this test?

azlev replied:
Oh, just saw here. The `filebeat/scripts/tester/main_test.go` I think that I've fixed, Now I'm looking at the alias.

jsoriano replied:
All green now :) thanks!

jsoriano reviewed (approved):
LGTM, let's wait for a green build. Thanks @azlev!

jsoriano commented:
jenkins run the tests

jsoriano commented:
/test

azlev commented:
Hey @jsoriano, I found a test that uses the postgresql ingest inside `tester`

I changed it and executed here:
```
BEAT_STRICT_PERMS=false mage build test
```

And I think that I fixed that error. Can you execute the pipeline again?

jsoriano commented on filebeat/module/postgresql/log/ingest/pipeline-csv.yml:
```diff
@@ -0,0 +1,78 @@
+    - '^duration: %{NUMBER:temp.duration:float} ms$'
+    - '^duration: %{NUMBER:temp.duration:float} ms  %{POSTGRESQL_QUERY_STEP:postgresql.log.query_step} %{DATA:postgresql.log.query_name}: %{GREEDYDATA:message}$'
+    - '^duration: %{NUMBER:temp.duration:float} ms  %{POSTGRESQL_QUERY_STEP:postgresql.log.query_step}: %{GREEDYDATA:message}$'
+    - '^(%{POSTGRESQL_QUERY_STEP}: )?%{GREEDYDATA:message}$'
```
Set the field also here?
```suggestion
    - '^(%{POSTGRESQL_QUERY_STEP:postgresql.log.query_step}: )?%{GREEDYDATA:message}$'
```

azlev replied:
Ops :-)

jsoriano commented on filebeat/module/postgresql/log/_meta/fields.yml:
```diff
@@ -28,25 +43,75 @@
+      description: >
+        A possible solution to solve an error.
+    - name: internal_query
+      description: internal query that led to the error (if any).
```
Please uppercase first letter of sentences in these descriptions.
```suggestion
      description: >
        Internal query that led to the error (if any).
```

jsoriano commented on filebeat/scripts/tester/main_test.go:
```diff
@@ -34,7 +34,7 @@ func TestGetPipelinePath(t *testing.T) {
 		{
 			pipelinePath: "../../module/postgresql/log/ingest",
-			count:        1,
+			count:        3,
```
:slightly_smiling_face:

jsoriano reviewed (commented):
Looks good. Only a couple of small details, sorry I didn't see them before.

azlev commented:
:eyes:

jsoriano reviewed (approved):
Great :+1:

jsoriano commented:
Merged and backported to 7.x, this will be hopefuly available in 7.12, thanks @azlev!
